engine/model_contact.go

```diff
@@ -0,0 +1,205 @@
+package engine
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"github.com/bitcoin-sv/go-paymail"
+	"github.com/bitcoin-sv/spv-wallet/engine/utils"
+	"github.com/mrz1836/go-cachestore"
+	"github.com/mrz1836/go-datastore"
+)
+
+type Contact struct {
+	// Base model
+	Model `bson:",inline"`
+
+	// Model specific fields
+	ID       string        `json:"id" toml:"id" yaml:"id" gorm:"<-:create;type:char(64);primaryKey;comment:This is the unique contact id" bson:"_id"`
+	XpubID   string        `json:"xpub_id" toml:"xpub_id" yaml:"xpub_id" gorm:"<-:create;type:char(64);foreignKey:XpubID;reference:ID;index;comment:This is the related xPub" bson:"xpub_id"`
+	FullName string        `json:"full_name" toml:"full_name" yaml:"full_name" gorm:"<-create;comment:This is the contact's full name" bson:"full_name"`
+	Paymail  string        `json:"paymail" toml:"paymail" yaml:"paymail" gorm:"<-create;comment:This is the paymail address alias@domain.com" bson:"paymail"`
+	PubKey   string        `json:"pub_key" toml:"pub_key" yaml:"pub_key" gorm:"<-:create;index;comment:This is the related public key" bson:"pub_key"`
+	Status   ContactStatus `json:"status" toml:"status" yaml:"status" gorm:"<-create;type:varchar(20);default:not authenticated;comment:This is the contact status" bson:"status"`
+}
+
+func newContact(fullName, paymailAddress, senderPubKey string, opts ...ModelOps) (*Contact, error) {
+	if fullName == "" {
+		return nil, ErrEmptyContactFullName
+	}
+
+	err := paymail.ValidatePaymail(paymailAddress)
+
+	if err != nil {
+		return nil, err
+	}
+
+	if senderPubKey == "" {
+		return nil, ErrEmptyContactPubKey
+	}
+
+	xPubId := utils.Hash(senderPubKey)
+
+	id := utils.Hash(xPubId + paymailAddress)
+
+	contact := &Contact{ID: id, XpubID: xPubId, Model: *NewBaseModel(ModelContact, opts...), FullName: fullName, Paymail: paymailAddress}
+
+	return contact, nil
+}
+
+func getContact(ctx context.Context, fullName, paymailAddress, senderPubKey string, opts ...ModelOps) (*Contact, error) {
+	contact, err := newContact(fullName, paymailAddress, senderPubKey, opts...)
+	if err != nil {
+		return nil, err
+	}
+	contact.ID = ""
+	conditions := map[string]interface{}{
```

Comment:
We should search for contact by sender xpub and contact paymail

Reply:
Changed